core/src/main/java/org/apache/calcite/rel/rules/SetOpToFilterRule.java

```diff
@@ -0,0 +1,188 @@
+ * Only inspect a single {@link Filter} layer in the inputs of {@link SetOp}.
+ * For inputs with nested {@link Filter}s, apply {@link CoreRules#FILTER_MERGE}
+ * as a preprocessing step.
+ *
+ * <p>Example:
+ *
+ * <p>UNION
+ * <blockquote><pre>
+ * SELECT mgr, comm FROM emp WHERE mgr = 12
+ * UNION
+ * SELECT mgr, comm FROM emp WHERE comm = 5
+ *
+ * to
+ *
+ * SELECT DISTINCT mgr, comm FROM emp
+ * WHERE mgr = 12 OR comm = 5
+ * </pre></blockquote>
+ *
+ * <p>INTERSECT
+ * <blockquote><pre>
+ * SELECT mgr, comm FROM emp WHERE mgr = 12
+ * INTERSECT
+ * SELECT mgr, comm FROM emp WHERE comm = 5
+ *
+ * to
+ *
+ * SELECT DISTINCT mgr, comm FROM emp
+ * WHERE mgr = 12 AND comm = 5
+ * </pre></blockquote>
+ *
+ * <p>EXCEPT
+ * <blockquote><pre>
+ * SELECT mgr, comm FROM emp WHERE mgr = 12
+ * EXCEPT
+ * SELECT mgr, comm FROM emp WHERE comm = 5
+ *
+ * to
+ *
+ * SELECT DISTINCT mgr, comm FROM emp
+ * WHERE mgr = 12 AND NOT(comm = 5)
+ * </pre></blockquote>
+ */
+@Value.Enclosing
+public class SetOpToFilterRule
+    extends RelRule<SetOpToFilterRule.Config>
+    implements TransformationRule {
+
+  /** Creates an SetOpToFilterRule. */
+  protected SetOpToFilterRule(Config config) {
+    super(config);
+  }
+
+  //~ Methods ----------------------------------------------------------------
+
+  @Override public void onMatch(RelOptRuleCall call) {
+    config.matchHandler().accept(this, call);
+  }
+
+  private void match(RelOptRuleCall call) {
+    final SetOp setOp = call.rel(0);
+    if (setOp.all || setOp.getInputs().size() != 2) {
+      return;
+    }
+    final RelBuilder builder = call.builder();
+    final RelNode leftInput = call.rel(1);
+    final Filter rightInput = call.rel(2);
+
+    if (!RexUtil.isDeterministic(rightInput.getCondition())) {
+      return;
+    }
+
+    RelNode leftBase;
+    RexNode leftCond = null;
+    if (leftInput instanceof Filter) {
+      Filter leftFilter = (Filter) leftInput;
+      leftBase = leftFilter.getInput().stripped();
+      leftCond = leftFilter.getCondition();
+    } else {
+      leftBase = leftInput.stripped();
+    }
+
+    final RelNode rightBase = rightInput.getInput().stripped();
+    if (!leftBase.equals(rightBase)) {
+      return;
+    }
+
+    RexNode finalCond = null;
+    // Right input is Filter, right cond should be not null
+    if (setOp instanceof Union) {
+      finalCond = leftCond != null
+          ? builder.or(leftCond, rightInput.getCondition())
+          : builder.literal(true);
+    } else if (setOp instanceof Intersect) {
+      finalCond = leftCond != null
+          ? builder.and(leftCond, rightInput.getCondition())
+          : rightInput.getCondition();
+    } else if (setOp instanceof Minus) {
+      finalCond = leftCond != null
+          ? builder.and(leftCond, builder.not(rightInput.getCondition()))
+          : builder.not(rightInput.getCondition());
+    } else {
+      // Should not enter here
```

Comment:
throw some exception that says "unreachable".
maybe there are other examples you can copy.